### PR TITLE
Better error message when method needs to be applied as partial function

### PR DESCRIPTION
**Pull request by kiritsuku**
The idea for a better error message occurred to me when I read the
following question on StackOverflow:

```
http://stackoverflow.com/questions/30846760
```

In short, the user did this:

```
scala> def yyy(c: Char) = {
     |     c.toUpper
     | }
    yyy: (c: Char)Char

scala> yyy('a')
res5: Char = A

scala> yyy
<console>:9: error: missing arguments for method yyy;
follow this method with `_' if you want to treat it as a partially
applied function
          yyy
          ^

scala> yyy_
<console>:8: error: not found: value yyy_
          yyy_
          ^
```

The current error message is clearly misleading, therefore a better one
is provided.

### CODE REVIEW

**Comment by som-snytt** [comment]
This is a good idea. I'd suggest a few more words. 

```
Methods are turned into function values only if the expected type is a Function. 
Or, you can write a partially applied function using the notation `f(_)' or `f _'.
```

**Comment by SethTisue** [comment]
A few thoughts.

First, this error message is so common, so ubiquitous, changing it for 2.11.x feels kinda weird to me. I recognize this error message on sight, at a glance, I've seen it so many times. How about targeting 2.12 instead...? (Or am I being absurdly fussy?)

Second... well, first of all, I'm glad you opened this PR, because I've never been that comfortable with the "helpfulness" of this error message. The "missing arguments for method yyy" part is of course fine, but how much basis does the compiler really have to suggest possible fixes? An expert user of the language already knows how and when to use underscore, so the hint is only potentially helpful for beginners, but I feel like it's as likely to lead a beginner astray as help them. And I understand the motivation for Som's suggested change, but it also makes the hint longer and more elaborate, and presents the beginner with a rather subtle and difficult to make choice.

Basically, I think an error message shouldn't try to teach you the language.

So, all of this is off the top of my head, but I wonder if it might actually be best to omit the hint entirely...? Or perhaps to add Som's "Methods are turned into function values only if the expected type is a Function" but get rid of the suggested fix.

Opinions welcome.

**Comment by adriaanm** [comment]
I think it's okay to change the second line, as preserving the first one is enough to maintain a connection to the status quo. It's also desirable to change it, as it's both unclear and outdated.

**Comment by adriaanm** [comment]
I'd suggest a variation on @som-snytt's suggestion:

```
Unapplied methods are only converted to functions when a function type is expected.
You may perform this conversion explicitly by writing `f(_)` or `f _` instead of `f`.
```

Using `${meth.name}` for `f`, as you did in this patch.

I'll leave it up to @SethTisue whether to postpone to 2.11.8.

**Comment by som-snytt** [comment]
+1 to Adriaan's wording. It'd be boss if `f(a, b, _)` reflected the application. The last polish is to make sure the first, original sentence has a period instead of a semi-colon. And I've loved semis (in prose) ever since childhood.

I agree with the sentiment that error messages aren't tutorials. Let's replace it with an error code.

Or better yet, instead of messages, emit a link to StackOverflow.
```
error: 42: http://stackoverflow.com/a/7890032/
```

**Comment by som-snytt** [comment]
Eleven days ago feels like an eternity. That's about a word a day, from the original PR. I got antsy and proposed another. I now feel that all trivial wording improvements should have equal opportunity to express themselves when the bug bites.

**Comment by kiritsuku** [comment]
@SethTisue Should I reopen the PR for 2.12 or just improve the error message?

**Comment by SethTisue** [comment]
let's take it to #4586
